File: Contest/1.cpp
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        vector<vector<int>> row(n, vector<int>(m, 0));
        int res = 0;

        // Step 1: build row prefix (number of consecutive 1s to the left)
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (mat[i][j] == 1) {
                    row[i][j] = (j == 0 ? 1 : row[i][j - 1] + 1);
                }
            }
        }

        // Step 2: count submatrices with (i, j) as bottom-right
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int cur = row[i][j];
                for (int k = i; k >= 0 && cur > 0; k--) {
                    cur = min(cur, row[k][j]);
                    res += cur;
                }
            }
        }

        return res;
    }
};

File: Contest/1.py
class Solution(object):
    def countSquares(self, matrix):
        row, col = len(matrix), len(matrix[0])
        dp = [[0] * col for _ in range(row)]

        count = 0
        for i in range(row):
            for j in range(col):
                if matrix[i][j] == 1:
                    if i == 0 or j == 0:
                        dp[i][j] = 1
                    else:
                        dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1
                    count += dp[i][j]
        return count